### PR TITLE
feat: add flush guard to PersistenceWindows

### PR DESCRIPTION
**Pull request by tustvold**
We have two requirements for persistence:

1. Don't stall ingest whilst persisting a partition
2. Don't flush data from PersistenceWindow till it is actually persisted

Criteria 1 means that we must drop the partition lock whilst persisting, but this means that the persistable window may be mutated whilst this is taking place. When we then flush the persistable window after having persisted, currently we may end up flushing more data than we actually persisted!

The simplest way around this is to stick a lock around the persistable window that is held whilst the persistence operation is running. Once persistence has completed, we re-acquire a write lock on the partition, drop the flush guard and then flush the persistable window. We are guaranteed that what is flushed hasn't changed since the start of the persistence action.

As the persistence operation is async the resulting lock guard must be `Send`. A tokio lock would work, but as we don't need to ever wait for a lock, `std::sync::Arc` provides all the functionality necessary.

re https://github.com/influxdata/conductor/issues/360

### CODE REVIEW

File: mutable_buffer/src/persistence_windows.rs [hunk 1]
```diff
@@ -126,38 +136,63 @@ impl PersistenceWindows {
...
+        );
+
+        let persistable = self
+            .persistable
+            .get_mut()
+            .expect("expected no active locks");
+
+        if let Some(persistable) = persistable {
+            // Everything up to and including persistable max time will have been persisted
+            let new_min = Utc.timestamp_nanos(persistable.max_time.timestamp_nanos() + 1);
```
Comment (tustvold) [inline mutable_buffer/src/persistence_windows.rs:183]: The split plan flushes <= and so the new min is actually 1 greater

File: internal_types/src/guard.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+/// let mut data = ReadLock::new(0);
+/// let mut_borrow = data.get_mut();
+///
+/// data.lock(); // Shouldn't compile
+///
+/// std::mem::drop(mut_borrow);
+/// ```
+///
+#[derive(Debug)]
+pub struct ReadLock<T>(Arc<T>);
```
Comment (mkmik) [inline internal_types/src/guard.rs:42]: nice

File: mutable_buffer/src/persistence_windows.rs [hunk 3]
```diff
@@ -17,15 +18,26 @@ const DEFAULT_CLOSED_WINDOW_PERIOD: Duration = Duration::from_secs(30);
...
 pub struct PersistenceWindows {
-    persistable: Option<Window>,
+    persistable: ReadLock<Option<Window>>,
     closed: VecDeque<Window>,
     open: Option<Window>,
     late_arrival_period: Duration,
     closed_window_period: Duration,
 }
 
+/// A handle for flushing data from the `PersistenceWindows`
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:30]:
```suggestion
/// A handle for flushing data from the `PersistenceWindows`
/// while preventing additional modification to the `persistable` list
```

File: mutable_buffer/src/persistence_windows.rs [hunk 4]
```diff
@@ -126,38 +136,63 @@ impl PersistenceWindows {
             self.closed.push_back(self.open.take().unwrap())
         }
 
-        while let Some(w) = self.closed.pop_front() {
-            if now.duration_since(w.created_at) >= self.late_arrival_period {
-                match self.persistable.as_mut() {
-                    Some(persistable_window) => persistable_window.add_window(w),
-                    None => self.persistable = Some(w),
+        if let Some(persistable) = self.persistable.get_mut() {
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:139]:
```suggestion
        // if there is no ongoing persistence operation, try and 
        // add closed windows to the `perstable` list
        if let Some(persistable) = self.persistable.get_mut() {
```

File: mutable_buffer/src/persistence_windows.rs [hunk 5]
```diff
@@ -126,38 +136,63 @@ impl PersistenceWindows {
...
 
-            if let Some(w) = self.open.as_mut() {
-                if w.min_time < t {
-                    w.min_time = t;
+    /// Clears out the persistable window
+    pub fn flush(&mut self, handle: FlushHandle) {
+        let closed_count = handle.closed_count;
+        std::mem::drop(handle);
+
+        assert!(
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:171]: 👍

File: mutable_buffer/src/persistence_windows.rs [hunk 6]
```diff
@@ -126,38 +136,63 @@ impl PersistenceWindows {
...
+            self.closed.len() >= closed_count,
+            "windows dropped from closed whilst locked"
+        );
+
+        let persistable = self
+            .persistable
+            .get_mut()
+            .expect("expected no active locks");
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:179]: 👍

File: mutable_buffer/src/persistence_windows.rs [hunk 7]
```diff
@@ -834,25 +881,196 @@ mod tests {
...
+        // Flush persistable window
+        w.flush(guard);
+        assert_eq!(w.persistable_row_count(), 0);
+
+        // This should rotate into persistable
+        w.rotate(instant + Duration::from_secs(240));
+        assert_eq!(w.persistable_row_count(), 5);
+
+        // Min time should have been truncated by persist operation to be
+        // o nanosecond more than was persisted
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:959]:
```suggestion
        // 3 nanosecond more than was persisted
```

File: mutable_buffer/src/persistence_windows.rs [hunk 8]
```diff
@@ -834,25 +881,196 @@ mod tests {
...
         let c = w.open.as_ref().unwrap();
         assert_eq!(c.row_count, 2);
-        assert_eq!(c.min_time, max_time);
+        assert_eq!(c.min_time, third_start);
         assert_eq!(c.max_time, third_end);
         assert_eq!(c.created_at, third_created_at);
     }
+
+    #[test]
+    fn test_flush_guard() {
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:908]: Nice tests

**Review by alamb — APPROVED** [review]
@tustvold gave me the backstory here

The usecase of the code in this PR is to prevent changes to the `persistable` list of `Window`s in the DB wide persistence windows while data is being persisted, but then relinquish that guard once the persistence action has completed.

File: mutable_buffer/src/persistence_windows.rs [hunk 9]
```diff
@@ -126,38 +136,63 @@ impl PersistenceWindows {
...
-    pub fn flush(&mut self) {
-        if let Some(t) = self.max_persistable_timestamp() {
-            for w in &mut self.closed {
-                if w.min_time < t {
-                    w.min_time = t;
-                }
-            }
+    /// Acquire a handle that prevents mutation of the persistable window until dropped
+    ///
+    /// Returns `None` if there is an outstanding read guard
```
Comment (alamb) [inline mutable_buffer/src/persistence_windows.rs:156]:
```suggestion
    /// Returns `None` if there is an outstanding handle
```

**Pull request merged**